### PR TITLE
Fixed up some initial logging support for instructor

### PR DESCRIPTION
**Pull request by ivanleomk**
| :rocket: | This description was created by [Ellipsis](https://www.ellipsis.dev) for commit b3f50a5e24b6d9ba3326bf9d78801fc8f2460197  | 
|--------|--------|

### Summary:
Added `logfire_api` logging support to various functions in the `instructor` module for enhanced tracing and error logging.

**Key points**:
- Added `logfire_api` for logging support in `instructor/client.py`, `instructor/patch.py`, `instructor/process_response.py`, and `instructor/retry.py`.
- Wrapped key function calls in `logfire.span` for tracing.
- Logged function results and errors using `logfire.info` and `logfire.error`.
- Updated `pyproject.toml` to include `logfire-api` dependency.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

**Comment by ivanleomk** [comment]
Currently what I've got it looking like is the following

CleanShot 2024-07-22 at 17 23 46@2x

This corresponds to the file below

```
class User(BaseModel):
    name: str
    age: int

    @field_validator("name")
    @classmethod
    def validate_uppercase(cls, v: str) -> str:
        if not v.isupper():
            raise ValueError("Uppercase Names only")
        return v


class Users(BaseModel):
    users: list[User]


client = instructor.from_openai(openai.OpenAI())


def get_user():
    resp = client.chat.completions.create(
        response_model=User,
        messages=[
            {
                "role": "user",
                "content": "Extract user data from this sentence - Ivan is a 27 year old developer from Singapore",
            }
        ],
        model="gpt-4o-mini",
    )

    print(resp)


get_user()
```

**Comment by ivanleomk** [comment]
CleanShot 2024-07-22 at 17 27 23@2x

This is what the async client will look like when we call it with the `.create` method.

**Comment by ivanleomk** [comment]
Some issues with getting Iterables to be logged. Will debug it

File: instructor/client.py [hunk 1]
```diff
@@ -153,19 +163,30 @@ def create_partial(
...
     ) -> Generator[T, None, None] | AsyncGenerator[T, None]:
-        kwargs["stream"] = True
-
-        kwargs = self.handle_kwargs(kwargs)
-
-        response_model = instructor.Partial[response_model]  # type: ignore
-        return self.create_fn(
+        with logfire.span(
+            "Creating Streaming Chat Completion",
+            response_model=response_model.model_json_schema()  # type:ignore
```
Comment (jxnl) [inline instructor/client.py:168]: this might large. could be an inssue

File: instructor/client.py [hunk 2]
```diff
@@ -153,19 +163,30 @@ def create_partial(
...
         **kwargs: Any,
     ) -> Generator[T, None, None] | AsyncGenerator[T, None]:
-        kwargs["stream"] = True
-
-        kwargs = self.handle_kwargs(kwargs)
-
-        response_model = instructor.Partial[response_model]  # type: ignore
-        return self.create_fn(
+        with logfire.span(
+            "Creating Streaming Chat Completion",
```
Comment (jxnl) [inline instructor/client.py:167]: snake case prefered

File: instructor/client.py [hunk 3]
```diff
@@ -109,16 +108,29 @@ def create(
         strict: bool = True,
         **kwargs: Any,
     ) -> T | Any | Awaitable[T] | Awaitable[Any]:
-        kwargs = self.handle_kwargs(kwargs)
-
-        return self.create_fn(
-            response_model=response_model,
+        with logfire.span(
+            "Creating Chat Completion",
```
Comment (jxnl) [inline instructor/client.py:112]: prefix with something so folks know its instructor

File: instructor/client.py [hunk 4]
```diff
@@ -294,15 +333,28 @@ async def create(
         strict: bool = True,
         **kwargs: Any,
     ) -> T | Any:
-        kwargs = self.handle_kwargs(kwargs)
-        return await self.create_fn(
-            response_model=response_model,
-            validation_context=validation_context,
-            max_retries=max_retries,
+        with logfire.span(
+            "Creating Async Chat Completion",
```
Comment (jxnl) [inline instructor/client.py:337]: what does this get that doesnt come from openai's instrumentation?

File: instructor/client.py [hunk 5]
```diff
@@ -75,8 +75,7 @@ def create(
         validation_context: dict[str, Any] | None = None,
         strict: bool = True,
         **kwargs: Any,
-    ) -> T:
-        ...
+    ) -> T: ...
```
Comment (jxnl) [inline instructor/client.py:78]: lots of linting?

**Pull request closed without merging**